pochivm/api_base.h

```diff
@@ -115,113 +121,161 @@ Value<T>::operator Value<U>() const
+          typename = std::enable_if_t<AstTypeHelper::primitive_type_supports_binary_op<T, AstTypeHelper::BinaryOps::MUL>::value> >
+Value<typename AstTypeHelper::ArithReturnType<T, U>::type> operator*(const Value<T>& lhs, const Value<U>& rhs)
 {
-    return Value<T>(new AstArithmeticExpr(AstArithmeticExprType::MUL, lhs.__pochivm_value_ptr, rhs.__pochivm_value_ptr));
+    return DoArith<AstArithmeticExprType::MUL>(lhs, rhs);
 }
 
-template<typename T, typename = std::enable_if_t<
-             AstTypeHelper::primitive_type_supports_binary_op<T, AstTypeHelper::BinaryOps::DIV>::value> >
-Value<T> operator/(const Value<T>& lhs, const Value<T>& rhs)
+template <typename T, typename U,
+          typename = std::enable_if_t<AstTypeHelper::primitive_type_supports_binary_op<U, AstTypeHelper::BinaryOps::DIV>::value>,
+          typename = std::enable_if_t<AstTypeHelper::primitive_type_supports_binary_op<T, AstTypeHelper::BinaryOps::DIV>::value> >
+Value<typename AstTypeHelper::ArithReturnType<T, U>::type> operator/(const Value<T>& lhs, const Value<U>& rhs)
 {
-    return Value<T>(new AstArithmeticExpr(AstArithmeticExprType::DIV, lhs.__pochivm_value_ptr, rhs.__pochivm_value_ptr));
+    return DoArith<AstArithmeticExprType::DIV>(lhs, rhs);
 }
 
-template<typename T, typename = std::enable_if_t<
-             AstTypeHelper::primitive_type_supports_binary_op<T, AstTypeHelper::BinaryOps::MODULO>::value> >
-Value<T> operator%(const Value<T>& lhs, const Value<T>& rhs)
+template <typename T, typename U,
+          typename = std::enable_if_t<AstTypeHelper::primitive_type_supports_binary_op<U, AstTypeHelper::BinaryOps::MODULO>::value>,
+          typename = std::enable_if_t<AstTypeHelper::primitive_type_supports_binary_op<T, AstTypeHelper::BinaryOps::MODULO>::value> >
+Value<typename AstTypeHelper::ArithReturnType<T, U>::type> operator%(const Value<T>& lhs, const Value<U>& rhs)
 {
-    return Value<T>(new AstArithmeticExpr(AstArithmeticExprType::MOD, lhs.__pochivm_value_ptr, rhs.__pochivm_value_ptr));
+    return DoArith<AstArithmeticExprType::MOD>(lhs, rhs);
 }
 
 // Convenience overloading: arithmetic operation with literal
 //
-template<typename T, typename = std::enable_if_t<
-             AstTypeHelper::primitive_type_supports_binary_op<T, AstTypeHelper::BinaryOps::ADD>::value> >
-Value<T> operator+(const Value<T>& lhs, T rhs)
+template <typename T, typename U,
+          typename = std::enable_if_t<AstTypeHelper::primitive_type_supports_binary_op<U, AstTypeHelper::BinaryOps::ADD>::value>,
+          typename = std::enable_if_t<AstTypeHelper::primitive_type_supports_binary_op<T, AstTypeHelper::BinaryOps::ADD>::value> >
+Value<typename AstTypeHelper::ArithReturnType<T, U>::type> operator+(const Value<T>& lhs, U rhs)
 {
-    return Value<T>(new AstArithmeticExpr(AstArithmeticExprType::ADD, lhs.__pochivm_value_ptr, new AstLiteralExpr(TypeId::Get<T>(), &rhs)));
+    return DoArith<AstArithmeticExprType::ADD>(lhs, Literal<U>(rhs));
```

Comment:
one issue with your current approach is that code like 'a+1' where a is int64_t would generate AST like Add(a, StaticCast<uint64_t>(Literal<int32_t>(1)), which is kind of bad. You can fix this by doing the rewrite in DoArith. (i.e. in DoArith, check if the side to be StaticCast'ed is a Literal, and if it is, directly cast its content).

Reply:
I'm not sure how I would directly cast a literal's content. Do you think I should add a GetAs<T>() method to the AstLiteralExpr class that returns a literal's value in the form of type T? Then I can cast that value to the desired type and put that in a new Literal? Alternatively, I could just go back to the old method with multiple overloads for ltierals on the left/right.

Reply:
yes that's definitely one way to do it.